Mathlib/CategoryTheory/Sites/LocalSite.lean

```diff
@@ -0,0 +1,145 @@
+/-
+Copyright (c) 2025 Ben Eltschig. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ben Eltschig
+-/
+import Mathlib.CategoryTheory.Adjunction.Triple
+import Mathlib.CategoryTheory.Sites.GlobalSections
+
+/-!
+# Local sites
+A site (C,J) is called local if it has a terminal object whose only covering sieve is trivial -
+this makes it possible to define codiscrete sheaves on it, giving its sheaf topos the structure
+of a local topos. This makes them an important stepping stone to cohesive sites.
+
+See https://ncatlab.org/nlab/show/local+site.
+
+## Main definitions / results
+* `LocalSite J`: typeclass stating that (C,J) is a local site.
+* `Sheaf.codisc`: the codiscrete functor `Type w ⥤ Sheaf J (Type max v w)` defined on any
+  local site.
+* `Sheaf.ΓCodiscAdj`: the adjunction between the global sections functor `Γ` and `Sheaf.codisc`.
+* `Sheaf.fullyFaithfulCodisc`: `Sheaf.codisc` is fully faithful.
+* `fullyFaithfulConstantSheaf`: on local sites, the constant sheaf functor is fully faithful.
+All together this shows that for local sites `Sheaf J (Type max u v w)` forms a local topos, but
+since we don't yet have local topoi this can't be stated yet.
+-/
+
+universe u v w
+
+open CategoryTheory Limits Sheaf Opposite GrothendieckTopology
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C] (J : GrothendieckTopology C)
+
+/-- A local site is a site that has a terminal object with only a single covering sieve. -/
+class LocalSite extends HasTerminal C where
```

Comment:
As this is a `Prop`, we usually like to use the prefix `Is`. This should probably go into the `GrothendieckTopology` namespace, so that we may write `J.IsLocal`, or `J.IsLocalSite`.